db/migrate/20170825144733_repopulate_first_published_at_from_change_history.rb

```diff
@@ -0,0 +1,39 @@
+class RepopulateFirstPublishedAtFromChangeHistory < ActiveRecord::Migration[5.1]
+  disable_ddl_transaction!
+
+  def up
+    min_date = '2016-02-29T09:24:09'
+    max_date = '2016-02-29T09:24:11'
+    document_ids = []
+
+    sql = <<-SQL.strip_heredoc
+      SELECT e.id,
+             e.document_id,
+             hist->>'public_timestamp'
+      FROM   editions e,
+             json_array_elements(details->'change_history') hist
+      WHERE  e.document_type != 'placeholder'
+      AND    e.first_published_at BETWEEN '#{min_date}' AND '#{max_date}'
+      AND    hist IS NOT NULL
+      AND    hist->>'note' = 'First published.'
+      AND    hist->>'public_timestamp' NOT BETWEEN '#{min_date}' AND '#{max_date}'
+    SQL
+
+    # ~224000 records
+    results = ActiveRecord::Base.connection.execute(sql).values
+    puts "#{results.size} records found with change history and first_published_at on 2016-02-29."
+
+    # ~395 seconds
+    results.each do |edition_id, document_id, timestamp|
+      sql = "UPDATE editions SET first_published_at = '#{timestamp}' WHERE id = #{edition_id};"
```

Comment:
If the slowness here is problematic, I wonder if we can make this faster by grouping lots of queries together with semi-colons into batches and then calling `execute` only a few times.

Reply:
:+1: I'll batch them up, this migration isn't blocking but I'm sure it can be run quicker than 6 minutes.